[workspace] Update macOS to python@3.12

MarcusSorealheis opened the pull request:
This PR Addresses the first task in this #20294 (switch Drake's default to 3.12)

If it looks ok and it lands, I can take on the external examples task as well. My company needs it for our project. The hope is to contribute more and more to Drake's maintenance and innovation over time since we benefit from it.

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/RobotLocomotion/drake/21013)
<!-- Reviewable:end -->

jwnimmer-tri commented:
Thanks @MarcusSorealheis!

+@BetsyMcPhail or +@svenevs or +@mwoehlke-kitware whoever has time, please sign up for feature review here (and un-assign the others).

+@jwnimmer-tri self-assigning for platform review.

Most of the work for this upgrade will be refreshing our CI machines, which will take some days.

jwnimmer-tri commented:
@drake-jenkins-bot ok to test

svenevs commented:
@MarcusSorealheis I did not do anything other than a quick pass, nominally things are looking good here!

Please compare with the diff in #18262 (our last macOS upgrade).  You're missing some things in the wheel build.  For the wheel files in particular, that one is an "unprovisioned build" meaning that the `setup/` scripts you are changing in this PR _will_ run.  When you have new code changes pushed up matching the additional files in there, post a GitHub (not reviewable) comment that says

```
@drake-jenkins-bot mac-arm-ventura-unprovisioned-clang-wheel-experimental-release please.
```

> Most of the work for this upgrade will be refreshing our CI machines, which will take some days.

Specifically, once we get the dust settled here, I'll be rebaking our "provisioned images" on the backend.  Until that is done, the pre-merge check mac-arm-ventura-clang-bazel-experimental-release will continue to fail on this PR (which is expected!  just letting you know).

Thanks for getting this started!

svenevs reviewed (commented):
-@BetsyMcPhail -@mwoehlke-kitware 

> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/21013#-:-NrGgYUD6qfsZ_t8Nyxm:b-wwepkw)* status: LGTM missing from assignees jwnimmer-tri(platform),svenevs



<!-- Sent from Reviewable.io -->

svenevs commented:
Also, re: wheel changes

https://github.com/RobotLocomotion/drake/blob/d899c0374629a22730a586ae015e005a0456473b/tools/wheel/wheel_builder/macos.py#L19-L22

when referring to #18262  you will not have instruction on this part.  The wheel gets built for both 3.11 and 3.12 right now, your PR should be changing this to just have 3.12 now:

```py
 python_targets = ( 
     PythonTarget(3, 12), 
 ) 
```

Down the road, when 3.13 is out, we'll add it to the list there (but the drake default will remain 3.12 until the next homebrew switch)

jwnimmer-tri commented:
@svenevs Actually can we keep building wheels for both 3.11 and 3.12 on macOS?  It seems like it would be possible, and of great value to our users.

svenevs commented:
@jwnimmer-tri Yes, I think we could do that, in which case the change to `python_targets` should **not** be done here.  However, I think it means we need to add python 3.11 to drake-ci explicitly (only) until we are in a place where we want to drop that.  I think this is the first switch like this where keeping the older macOS wheel is even possible.

svenevs commented:
Ok, sorry for the confusion, #20616 changed things a lot more than I realized.  We should discuss in f2f later today

jwnimmer-tri commented:
@drake-jenkins-bot mac-arm-ventura-unprovisioned-clang-bazel-experimental-release please

svenevs commented:
Hi @MarcusSorealheis, it looks like we're going to need to solve #8392 and possibly some other bits of CI changes before we're going to be able to make progress on this.  High level overview is homebrew disabled using `pip` to install directly into site-packages for python 3.12+, so we're going to concoct some kind of drake-specific likely virtualenv related change first.

Thanks for helping get this started so quickly (and therefore helping us identify the need to fix some other parts)!

MarcusSorealheis commented:
> High level overview is homebrew disabled using pip to install directly into site-packages for python 3.12+, so we're going to concoct some kind of drake-specific likely virtualenv related change first.

Ah, that's a bummer but positive overall. I am happy to help with this in any way if you want or need. This project's vitality is very important to me. At a minimum, tonight I will get up to speed on that issue from 2018 and then test a few options out. If I get interesting data, I will share it as well.  Thanks for being helpful.

jwnimmer-tri commented:
Maybe we're overthinking this.  Using a `venv` is absolutely the right long-term plan, but I wonder if just tacking on `--user` to all of the pip commands in our `setup/mac/...` scripts would solve the problem?

MarcusSorealheis commented:
> Maybe we're overthinking this. Using a venv is absolutely the right long-term plan, but I wonder if just tacking on --user to all of the pip commands in our setup/mac/... scripts would solve the problem?

I'm investigating now, but I made the change. I need to run it on my Mac.

MarcusSorealheis commented:
I'll need to schedule some time to dig into this issue tomorrow if someone doesn't beat me to it. A cursory look into the logs here shows a problem I am not seeing on my local Mac, as far as I can tell. I am on a different version than the CI.

MarcusSorealheis commented:
Is there anything else to do here @jwnimmer-tri besides reflashing the images?

jwnimmer-tri commented:
The build that we _do_ need to be passing right now (`mac-arm-ventura-unprovisioned-clang-bazel-experimental-release`) did not pass.  It looks at first glance like the `--user` flag wasn't a simple fix.

The bottom line is this: CI is just running setup scripts that we have in git.  Our users will run the same scripts.  Presumably, since they are failing in CI they will also fail for you when you run them on your local computer.  Until `install_prereqs` runs successfully locally (and you can do a full Drake `bazel test //...` afterwards), CI is just going to tell you the same things you already know -- the code is not working yet.

MarcusSorealheis commented:
@drake-jenkins-bot mac-arm-ventura-unprovisioned-clang-bazel-experimental-release please

MarcusSorealheis commented:
@jwnimmer-tri I read the related PEP ([0668](https://peps.python.org/pep-0668/)) and also the [updated canonical spec](https://packaging.python.org/en/latest/specifications/externally-managed-environments/#externally-managed-environments) and it seems that the latest update in this branch, provided it passes, should work. It also feels very clear that we should move off of it soon, and I will be happy to lend a hand with the migration to `venv`, `pipx`, or whatever environment isolation library is chosen. I could even write up an update to the situation in the existing issue [8392](https://github.com/RobotLocomotion/drake/issues/8392) from 2018. An excerpt from the spec makes the urgency clear.

> Packaging pipx in the distro avoids the irony of instructing users to pip install --user --break-system-packages pipx to _avoid_ breaking system packages.


However, I do consider this PR a step forward and something that should happen prior to virtualization.

jwnimmer-tri commented:
Hmm, possibly the bot commands need special permissions?  Anyway, I'll launch it:

@drake-jenkins-bot mac-arm-ventura-unprovisioned-clang-bazel-experimental-release please

MarcusSorealheis commented:
Ahh sorry @jwnimmer-tri the most recent failure was a known user error on my part. I've fixed it and things should work now 🤞 

The source distribution's version of `install_prereqs.sh` had two invocations of pip and both in that file were wrong. I've fixed that.

jwnimmer-tri commented:
@drake-jenkins-bot mac-arm-ventura-unprovisioned-clang-bazel-experimental-release please

jwnimmer-tri reviewed (commented):
This seems in pretty good shape so far.

I'm surveying people about whether the `--break-system-packages` flag will be a problem for them.

> Reviewed 6 of 8 files at r1, 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/21013#-:-Nrmvu1c4gksnaITpmhf:b-7cf1he)* status: 1 unresolved discussion, LGTM missing from assignees jwnimmer-tri(platform),svenevs

*[a discussion](https://reviewable.io/reviews/RobotLocomotion/drake/21013#-Nrmuh744nFSlGCw-mvJ:-Nrmuh744nFSlGCw-mvK:bpybxvv) (no related file):*
The CI shows one failure: it looks like a package we use (https://github.com/xmlrunner/unittest-xml-reporting) is incompatible with python3.12 and needs a patch, something along these lines:

```patch
--- xmlrunner/unittest.py
+++ xmlrunner/unittest.py
@@ -4,8 +4,9 @@
 import sys
 # pylint: disable-msg=W0611
 import unittest
-from unittest import TextTestRunner
-from unittest import TestResult, _TextTestResult
+from unittest import TextTestRunnerf
+from unittest import TestResult
+from unittest import TextTestResult as _TextTestResult
 from unittest.result import failfast
 from unittest.main import TestProgram
```

That _content_ would go into a file under https://github.com/RobotLocomotion/drake/tree/master/tools/workspace/xmlrunner_py named e.g. `patches/test_result.patch`, with something like this change to the neighboring `repository.bzl` file to cite it:

```patch
--- tools/workspace/xmlrunner_py/repository.bzl
+++ tools/workspace/xmlrunner_py/repository.bzl
@@ -9,5 +9,8 @@ def xmlrunner_py_repository(
         commit = "3.0.2",
         sha256 = "dbe165386952ec5373d4db5b4ac0644b60b734f4b02b9e575b1d0dc873616ba4",  # noqa
         build_file = ":package.BUILD.bazel",
+        patches = [
+            ":patches/test_result.patch",
+        ],
         mirrors = mirrors,
     )
```

The code in https://github.com/RobotLocomotion/drake/tree/master/tools/workspace/mypy_internal would be a good demo for how to patch things.

(I would try to push this fix myself, but I don't have a mac handy to test it on, so I'm not sure if the spelling is exactly right.)

___


<!-- Sent from Reviewable.io -->

jwnimmer-tri reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/21013)* status: 1 unresolved discussion, LGTM missing from assignees jwnimmer-tri(platform),svenevs

*[a discussion](https://reviewable.io/reviews/RobotLocomotion/drake/21013#-Nrmuh744nFSlGCw-mvJ:-Nrmwf8B34iZMKkTGt9p:b-m74os0) (no related file):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

The CI shows one failure: it looks like a package we use (https://github.com/xmlrunner/unittest-xml-reporting) is incompatible with python3.12 and needs a patch, something along these lines:

```patch
--- xmlrunner/unittest.py
+++ xmlrunner/unittest.py
@@ -4,8 +4,9 @@
 import sys
 # pylint: disable-msg=W0611
 import unittest
-from unittest import TextTestRunner
-from unittest import TestResult, _TextTestResult
+from unittest import TextTestRunnerf
+from unittest import TestResult
+from unittest import TextTestResult as _TextTestResult
 from unittest.result import failfast
 from unittest.main import TestProgram
```

That _content_ would go into a file under https://github.com/RobotLocomotion/drake/tree/master/tools/workspace/xmlrunner_py named e.g. `patches/test_result.patch`, with something like this change to the neighboring `repository.bzl` file to cite it:

```patch
--- tools/workspace/xmlrunner_py/repository.bzl
+++ tools/workspace/xmlrunner_py/repository.bzl
@@ -9,5 +9,8 @@ def xmlrunner_py_repository(
         commit = "3.0.2",
         sha256 = "dbe165386952ec5373d4db5b4ac0644b60b734f4b02b9e575b1d0dc873616ba4",  # noqa
         build_file = ":package.BUILD.bazel",
+        patches = [
+            ":patches/test_result.patch",
+        ],
         mirrors = mirrors,
     )
```

The code in https://github.com/RobotLocomotion/drake/tree/master/tools/workspace/mypy_internal would be a good demo for how to patch things.

(I would try to push this fix myself, but I don't have a mac handy to test it on, so I'm not sure if the spelling is exactly right.)
</blockquote></details>

Oops, indeed I had a typo in some of my suggestions above.  The only diff we want should be to fix the spelling of `_TextTestResult` vs `TextTestResult`.

___


<!-- Sent from Reviewable.io -->

MarcusSorealheis commented:
ok, I'll check this out. quickly putting out another fire. thanks for the help here.

jwnimmer-tri reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/21013)* status: 1 unresolved discussion, LGTM missing from assignees jwnimmer-tri(platform),svenevs

*[a discussion](https://reviewable.io/reviews/RobotLocomotion/drake/21013#-Nrmuh744nFSlGCw-mvJ:-Nrn6tyZ-WNEIVshAwvC:b-c7kiwq) (no related file):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Oops, indeed I had a typo in some of my suggestions above.  The only diff we want should be to fix the spelling of `_TextTestResult` vs `TextTestResult`.
</blockquote></details>

Ah, oops; no patching necessary, actually.

There is a fixed version `3.2.0` available, it's just not tagged in GitHub so it didn't show up for me.  In a separate PR, I'll merge that upgrade so we can rely on it here.

___


<!-- Sent from Reviewable.io -->

jwnimmer-tri reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/21013)* status: LGTM missing from assignees jwnimmer-tri(platform),svenevs

*[a discussion](https://reviewable.io/reviews/RobotLocomotion/drake/21013#-Nrmuh744nFSlGCw-mvJ:-Nrn8YH_Cp75l-R4OFql:burmzuc) (no related file):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Ah, oops; no patching necessary, actually.

There is a fixed version `3.2.0` available, it's just not tagged in GitHub so it didn't show up for me.  In a separate PR, I'll merge that upgrade so we can rely on it here.
</blockquote></details>

See #21071.

___


<!-- Sent from Reviewable.io -->

MarcusSorealheis reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/21013)* status: LGTM missing from assignees jwnimmer-tri(platform),svenevs

*[a discussion](https://reviewable.io/reviews/RobotLocomotion/drake/21013#-Nrmuh744nFSlGCw-mvJ:-Nrn8JkRDuTq-zk-8xwH:bjh4bdp) (no related file):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

See #21071.
</blockquote></details>

Wow. Phew. Thank you for all the hard work faster than I could get to the patchwork. 

___


<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
@drake-jenkins-bot mac-arm-ventura-unprovisioned-clang-bazel-experimental-release please

mwoehlke-kitware reviewed (commented):
> Reviewed 5 of 8 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/21013)* status: LGTM missing from assignees jwnimmer-tri(platform),svenevs



<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
I explained f2f today to @svenevs how to fix the `cpplint_unittest` error.  It will happen in a separate PR, and then we can re-test here again.

I don't have any insight on the `jupyter` unit test error.  (I haven't really investigated it.)

MarcusSorealheis commented:
I can investigate the Jupyter error tonight.



On Thu, Feb 29, 2024 at 12:35 Jeremy Nimmer ***@***.***>
wrote:

> I explained f2f today to @svenevs <https://github.com/svenevs> how to fix
> the cpplint_unittest error. It will happen in a separate PR, and then we
> can re-test here again.
>
> I don't have any insight on the jupyter unit test error. (I haven't
> really investigated it.)
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/RobotLocomotion/drake/pull/21013#issuecomment-1971914672>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAR6TSEGICQMZFA2DX5VBELYV6IKDAVCNFSM6AAAAABDUA5DVOVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTSNZRHEYTINRXGI>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

svenevs commented:
Ok great thank you! I am rearranging my schedule tomorrow right now and should be able to help you resolve anything left outstanding 🙂

MarcusSorealheis commented:
@jwnimmer-tri and @svenevs a bit more good fortune for this PR. The failing `jupyter` test appears to be related to the `XMLRunner` dependency in my first look into the problem. Here's what a bit of the console output in Jenkins looks like:

```shell
==================== Test output for //tools/jupyter:py/failing_notebook_test:
[6:19:36 PM]  Traceback (most recent call last):
[6:19:36 PM]    File "/Users/admin/workspace/mac-arm-ventura-unprovisioned-clang-bazel-experimental-release/_bazel_admin/4c3a1e08493e1ec72f883b9e38f87cf5/sandbox/darwin-sandbox/16693/execroot/drake/bazel-out/darwin_arm64-opt/bin/tools/jupyter/py/failing_notebook_test.runfiles/drake/common/test_utilities/drake_py_unittest_main.py", line 16, in <module>
[6:19:36 PM]      import xmlrunner
[6:19:36 PM]    File "/Users/admin/workspace/mac-arm-ventura-unprovisioned-clang-bazel-experimental-release/_bazel_admin/4c3a1e08493e1ec72f883b9e38f87cf5/sandbox/darwin-sandbox/16693/execroot/drake/bazel-out/darwin_arm64-opt/bin/tools/jupyter/py/failing_notebook_test.runfiles/xmlrunner_py/xmlrunner/__init__.py", line 11, in <module>
[6:19:36 PM]      from .runner import XMLTestRunner
[6:19:36 PM]    File "/Users/admin/workspace/mac-arm-ventura-unprovisioned-clang-bazel-experimental-release/_bazel_admin/4c3a1e08493e1ec72f883b9e38f87cf5/sandbox/darwin-sandbox/16693/execroot/drake/bazel-out/darwin_arm64-opt/bin/tools/jupyter/py/failing_notebook_test.runfiles/xmlrunner_py/xmlrunner/runner.py", line 6, in <module>
```

The first file and line referenced is [drake_py_unittest_main.py:16](https://github.com/RobotLocomotion/drake/blob/9faddc40891e2b63ae11b67f597286a6e8bce21d/common/test_utilities/drake_py_unittest_main.py#L16) and here's line 16 from that file:

```py
import xmlrunner
```
In that case, you don't need to investigate that one just yet. It is likely solved by #21071. That leaves only the `cpplint_unittest` error as the only remaining _known_ error. Is there already an issue or PR for tracking? This PR context is getting long 😅 , thanks for all the help again.

jwnimmer-tri commented:
The most recent CI log from `mac-arm-ventura-unprovisioned-clang-bazel-experimental-release` ([30](https://drake-jenkins.csail.mit.edu/job/mac-arm-ventura-unprovisioned-clang-bazel-experimental-release/30/)) is _already_ rebased on #21071.  The jupyter failure is https://drake-cdash.csail.mit.edu/test/1386038971 (as shown in that log), not an xmlrunner thing.

MarcusSorealheis commented:
> The most recent CI log from mac-arm-ventura-unprovisioned-clang-bazel-experimental-release ([30](https://drake-jenkins.csail.mit.edu/job/mac-arm-ventura-unprovisioned-clang-bazel-experimental-release/30/)) is already rebased ... The jupyter failure is ... not an xmlrunner thing.

Darn. I was looking at build [29](https://drake-jenkins.csail.mit.edu/job/mac-arm-ventura-unprovisioned-clang-bazel-experimental-release/29/) as I already had that tab open. I'll need to sink my teeth in after all.

MarcusSorealheis commented:
Looks like this one will take more than 15 minutes of digging. I'll get back to it later tonight.

jwnimmer-tri reviewed (commented):
FYi from my end -- I've finished cross-checking everything that I need to here now, so <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> for platform review once Jenkins is happy.

Remember that we'll need to eventually check both arm and x86 unprovisioned builds before merging this.  Also the Kitware folks will need to snapshot new images based on this code, and deploy those concurrently with merging this.

> Reviewed 1 of 2 files at r3.
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/21013#-:-NruooWN-MCsIjnDnCf5:bbz60ze)* status: LGTM missing from assignee svenevs



<!-- Sent from Reviewable.io -->

svenevs commented:
I was able to get things setup for the `styleguide` fix, but it appears to be more than just `assertEquals` => `assertEqual`, seems like they were possibly relying on some undocumented features

```py3tb
Traceback (most recent call last):
  File "/private/var/tmp/_bazel_stephen.mcdowell/c82709baf62040b4ad09f16721f3b518/sandbox/darwin-sandbox/23307/execroot/drake/bazel-out/darwin_arm64-opt/bin/tools/workspace/styleguide/cpplint_unittest.runfiles/styleguide/cpplint/cpplint_unittest.py", line 3063, in testTwoSpacesBetweenCodeAndComments
    self.TestLint('} // namespace foo',
  File "/private/var/tmp/_bazel_stephen.mcdowell/c82709baf62040b4ad09f16721f3b518/sandbox/darwin-sandbox/23307/execroot/drake/bazel-out/darwin_arm64-opt/bin/tools/workspace/styleguide/cpplint_unittest.runfiles/styleguide/cpplint/cpplint_unittest.py", line 224, in TestLint
    self.assertEqual(expected_message, self.PerformSingleLineLint(code))
                                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/private/var/tmp/_bazel_stephen.mcdowell/c82709baf62040b4ad09f16721f3b518/sandbox/darwin-sandbox/23307/execroot/drake/bazel-out/darwin_arm64-opt/bin/tools/workspace/styleguide/cpplint_unittest.runfiles/styleguide/cpplint/cpplint_unittest.py", line 129, in PerformSingleLineLint
    error_collector = ErrorCollector(self.assert_)
                                     ^^^^^^^^^^^^
AttributeError: 'CpplintTest' object has no attribute 'assert_'. Did you mean: 'assertIn'?
```

Will take a look again after some meetings.  @jwnimmer-tri you mentioned there might be a community fork somewhere?  If so, can you share a link?  Somebody has probably already solved this.  The `assert_` appears to be used as a lambda, will be switching to a different mac that's still setup with python 3.11 to triage soon.

jwnimmer-tri commented:
The other fork is https://github.com/cpplint/cpplint.  Looks like https://github.com/cpplint/cpplint/pull/182 is what you want.

svenevs commented:
@drake-jenkins-bot mac-arm-ventura-unprovisioned-clang-bazel-experimental-release please

svenevs reviewed (commented):
> Reviewed 5 of 8 files at r1, 1 of 2 files at r3, 1 of 1 files at r4, 1 of 1 files at r5, 1 of 1 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/21013)* status: 1 unresolved discussion, LGTM missing from assignee svenevs, commits need curation (https://drake.mit.edu/reviewable.html#curated-commits) (waiting on @MarcusSorealheis)

___
*[`tools/workspace/styleguide/repository.bzl` line 11 at r6](https://reviewable.io/reviews/RobotLocomotion/drake/21013#-NrvZsgM9iBA8rDnL3vd:-NrvZsgM9iBA8rDnL3ve:b-11dizg) ([raw file](https://github.com/RobotLocomotion/drake/blob/fdc228d1f861ed9b0037a6ac0bfd7a639e4f90ab/tools/workspace/styleguide/repository.bzl#L11)):*
> ```Starlark
>         # commit = "6aae49e2b5861cd298d681d628830a6261b947c3",
>         # sha256 = "4a46dedf35d61c0dfeeb4e3ce6c6593c59265624332d849268a787ca1cc95a2a",  # noqa
>         # TODO(svenevs): after a styleguide PR lands, do a drake PR to just
> ```

working, new PR in bound

@MarcusSorealheis FYI I had to rebase and force-push here, your local branch will be out of sync with what's on your fork's branch


<!-- Sent from Reviewable.io -->

svenevs reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/21013)* status: 1 unresolved discussion, LGTM missing from assignee svenevs, commits need curation (https://drake.mit.edu/reviewable.html#curated-commits) (waiting on @MarcusSorealheis)

___
*[`tools/workspace/styleguide/repository.bzl` line 11 at r6](https://reviewable.io/reviews/RobotLocomotion/drake/21013#-NrvZsgM9iBA8rDnL3vd:-Nrva48O9xOHuXPJ2sd9:b-irvde1) ([raw file](https://github.com/RobotLocomotion/drake/blob/fdc228d1f861ed9b0037a6ac0bfd7a639e4f90ab/tools/workspace/styleguide/repository.bzl#L11)):*
<details><summary><i>Previously, svenevs (Stephen McDowell) wrote…</i></summary><blockquote>

working, new PR in bound

@MarcusSorealheis FYI I had to rebase and force-push here, your local branch will be out of sync with what's on your fork's branch
</blockquote></details>

Step 1 is up: https://github.com/RobotLocomotion/styleguide/pull/54

@jwnimmer-tri I cannot assign reviewers there, please assign yourself if you have bandwidth.  Otherwise I'll request on slack.


<!-- Sent from Reviewable.io -->

svenevs commented:
@drake-jenkins-bot mac-arm-ventura-unprovisioned-clang-bazel-experimental-release please

this one may actually pass!

svenevs reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/21013)* status: 2 unresolved discussions, LGTM missing from assignee svenevs, commits need curation (https://drake.mit.edu/reviewable.html#curated-commits) (waiting on @MarcusSorealheis)

___
*[`tools/jupyter/test/jupyter_notebook_test.py` line 10 at r7](https://reviewable.io/reviews/RobotLocomotion/drake/21013#-NrvzV567p9vqdr9EWUO:-NrvzV567p9vqdr9EWUP:bxxnxo6) ([raw file](https://github.com/RobotLocomotion/drake/blob/28f2155512a70db4affc9aa7f03567a7fa2e2825/tools/jupyter/test/jupyter_notebook_test.py#L10)):*
> ```Python
>         """Ensures that `jupyter notebook` is installed (#12042)."""
>         status = subprocess.call(
>             args=[sys.executable, "-m", "notebook", "--help"])
> ```

working, waiting on test results, from commit message

2. We appear to be installing the wrong PyPI package!
  
    - [https://jupyter.org/install#jupyter-notebook](https://jupyter.org/install#jupyter-notebook)
    - [https://pypi.org/project/jupyter/](https://pypi.org/project/jupyter/) (very old...)
    - [https://pypi.org/project/notebook/](https://pypi.org/project/notebook/) (what we want...)

should be what we want now


<!-- Sent from Reviewable.io -->

svenevs reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/21013)* status: 3 unresolved discussions, LGTM missing from assignee svenevs, commits need curation (https://drake.mit.edu/reviewable.html#curated-commits) (waiting on @MarcusSorealheis)

___
*[`setup/mac/source_distribution/requirements-test-only.txt` line 2 at r7](https://reviewable.io/reviews/RobotLocomotion/drake/21013#-NrvzzxO9IX1marLXxHf:-Nrw-n0e1JZ9zv5_Nxzt:b-aynibp) ([raw file](https://github.com/RobotLocomotion/drake/blob/28f2155512a70db4affc9aa7f03567a7fa2e2825/setup/mac/source_distribution/requirements-test-only.txt#L2)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

nit This one is already in the binary prereqs `requirements.txt`.  So if this is really the one we want, then we can just drop this line.

It's fine to just fix that in the separate, ahead-of-time PR where we land this patch on its own, on master.
</blockquote></details>

gah! right you are... is there any reason that this PR cannot simply remove it?


<!-- Sent from Reviewable.io -->

MarcusSorealheis reviewed (commented):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 6 of 8 files at r1, 1 of 2 files at r3, 1 of 1 files at r4.
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/21013#-:-Nrw0Nd35ovjXxuAihPa:bnfp4nl)* status: 3 unresolved discussions, LGTM missing from assignee svenevs, commits need curation (https://drake.mit.edu/reviewable.html#curated-commits)



<!-- Sent from Reviewable.io -->

jwnimmer-tri reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/21013)* status: 3 unresolved discussions, LGTM missing from assignee svenevs, commits need curation (https://drake.mit.edu/reviewable.html#curated-commits) (waiting on @MarcusSorealheis)

___
*[`setup/mac/source_distribution/requirements-test-only.txt` line 2 at r7](https://reviewable.io/reviews/RobotLocomotion/drake/21013#-NrvzzxO9IX1marLXxHf:-Nrw0nojCTlCySGk5vS_:bp9htot) ([raw file](https://github.com/RobotLocomotion/drake/blob/28f2155512a70db4affc9aa7f03567a7fa2e2825/setup/mac/source_distribution/requirements-test-only.txt#L2)):*
<details><summary><i>Previously, svenevs (Stephen McDowell) wrote…</i></summary><blockquote>

gah! right you are... is there any reason that this PR cannot simply remove it?
</blockquote></details>

I want to land all of the "be compatible with python 3.12" as separate PRs ahead of time.  The "switch from 3.11 to 3.12" should only be switching, not any fixing.  That keeps the concerns separated, and makes revert wars and git history easier to work with.


<!-- Sent from Reviewable.io -->

svenevs reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/21013)* status: 3 unresolved discussions, LGTM missing from assignee svenevs, commits need curation (https://drake.mit.edu/reviewable.html#curated-commits) (waiting on @MarcusSorealheis)

___
*[`setup/mac/source_distribution/requirements-test-only.txt` line 2 at r7](https://reviewable.io/reviews/RobotLocomotion/drake/21013#-NrvzzxO9IX1marLXxHf:-Nrw2EalDUIaUZFaQ7na:b-btr01z) ([raw file](https://github.com/RobotLocomotion/drake/blob/28f2155512a70db4affc9aa7f03567a7fa2e2825/setup/mac/source_distribution/requirements-test-only.txt#L2)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

I want to land all of the "be compatible with python 3.12" as separate PRs ahead of time.  The "switch from 3.11 to 3.12" should only be switching, not any fixing.  That keeps the concerns separated, and makes revert wars and git history easier to work with.
</blockquote></details>

Good plan, future us will be thankful :wink: => #21085


<!-- Sent from Reviewable.io -->

svenevs reviewed (commented):
I went ahead and cancelled the CI here, looks like we're one rebase away :sunglasses: (jupyter fix)

> Reviewed 3 of 3 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/21013#-:-NrwHLUV22mpXoDcF_-0:b-apc86u)* status: 1 unresolved discussion, LGTM missing from assignee svenevs

___
*[`tools/jupyter/test/jupyter_notebook_test.py` line 10 at r7](https://reviewable.io/reviews/RobotLocomotion/drake/21013#-NrvzV567p9vqdr9EWUO:-NrwHAS782AcNeI_cgfJ:b-wx19gj) ([raw file](https://github.com/RobotLocomotion/drake/blob/28f2155512a70db4affc9aa7f03567a7fa2e2825/tools/jupyter/test/jupyter_notebook_test.py#L10)):*
<details><summary><i>Previously, svenevs (Stephen McDowell) wrote…</i></summary><blockquote>

working, waiting on test results, from commit message

2. We appear to be installing the wrong PyPI package!
  
    - [https://jupyter.org/install#jupyter-notebook](https://jupyter.org/install#jupyter-notebook)
    - [https://pypi.org/project/jupyter/](https://pypi.org/project/jupyter/) (very old...)
    - [https://pypi.org/project/notebook/](https://pypi.org/project/notebook/) (what we want...)

should be what we want now
</blockquote></details>

working, thread block until #21085 lands and can rebase here.
___
*[`tools/workspace/styleguide/repository.bzl` line 11 at r6](https://reviewable.io/reviews/RobotLocomotion/drake/21013#-NrvZsgM9iBA8rDnL3vd:-NrwHH828gjvTqCn_QP7:berbfzh) ([raw file](https://github.com/RobotLocomotion/drake/blob/fdc228d1f861ed9b0037a6ac0bfd7a639e4f90ab/tools/workspace/styleguide/repository.bzl#L11)):*
<details><summary><i>Previously, svenevs (Stephen McDowell) wrote…</i></summary><blockquote>

Step 1 is up: https://github.com/RobotLocomotion/styleguide/pull/54

@jwnimmer-tri I cannot assign reviewers there, please assign yourself if you have bandwidth.  Otherwise I'll request on slack.
</blockquote></details>

done, styleguide rebase is in here now


<!-- Sent from Reviewable.io -->

svenevs commented:
@drake-jenkins-bot linux-jammy-unprovisioned-clang-bazel-experimental-debug please
@drake-jenkins-bot linux-jammy-unprovisioned-clang-bazel-experimental-everything-debug please
@drake-jenkins-bot linux-jammy-unprovisioned-clang-bazel-experimental-everything-release please
@drake-jenkins-bot linux-jammy-unprovisioned-clang-bazel-experimental-release please
@drake-jenkins-bot linux-jammy-unprovisioned-gcc-bazel-experimental-debug please
@drake-jenkins-bot linux-jammy-unprovisioned-gcc-bazel-experimental-documentation please
@drake-jenkins-bot linux-jammy-unprovisioned-gcc-bazel-experimental-documentation please
@drake-jenkins-bot linux-jammy-unprovisioned-gcc-bazel-experimental-everything-debug please
@drake-jenkins-bot linux-jammy-unprovisioned-gcc-bazel-experimental-everything-release please
@drake-jenkins-bot linux-jammy-unprovisioned-gcc-bazel-experimental-mirror-to-s3 please
@drake-jenkins-bot linux-jammy-unprovisioned-gcc-bazel-experimental-packaging please
@drake-jenkins-bot linux-jammy-unprovisioned-gcc-bazel-experimental-packaging please
@drake-jenkins-bot linux-jammy-unprovisioned-gcc-bazel-experimental-release please
@drake-jenkins-bot linux-jammy-unprovisioned-gcc-bazel-experimental-release please
@drake-jenkins-bot linux-jammy-unprovisioned-gcc-cmake-experimental-debug please
@drake-jenkins-bot linux-jammy-unprovisioned-gcc-cmake-experimental-everything-debug please
@drake-jenkins-bot linux-jammy-unprovisioned-gcc-cmake-experimental-everything-release please
@drake-jenkins-bot linux-jammy-unprovisioned-gcc-cmake-experimental-release please
@drake-jenkins-bot linux-jammy-unprovisioned-gcc-wheel-experimental-release please
@drake-jenkins-bot mac-arm-sonoma-unprovisioned-clang-bazel-experimental-release please
@drake-jenkins-bot mac-arm-ventura-unprovisioned-clang-bazel-experimental-packaging please
@drake-jenkins-bot mac-arm-ventura-unprovisioned-clang-bazel-experimental-release please
@drake-jenkins-bot mac-arm-ventura-unprovisioned-clang-wheel-experimental-release please
@drake-jenkins-bot mac-x86-monterey-unprovisioned-clang-bazel-experimental-packaging please
@drake-jenkins-bot mac-x86-monterey-unprovisioned-clang-bazel-experimental-release please
@drake-jenkins-bot mac-x86-monterey-unprovisioned-clang-wheel-experimental-release please

svenevs reviewed (commented):
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> pending ci

> Reviewed 3 of 3 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/21013#-:-NrwQSl36W1VaHqjggqo:b-hwk7jy)* status: <span>:shipit:</span> complete! all discussions resolved, LGTM from assignees jwnimmer-tri(platform),svenevs

___
*[`tools/jupyter/test/jupyter_notebook_test.py` line 10 at r7](https://reviewable.io/reviews/RobotLocomotion/drake/21013#-NrvzV567p9vqdr9EWUO:-NrwJFR2Bbsrtl-kLkxc:b709npx) ([raw file](https://github.com/RobotLocomotion/drake/blob/28f2155512a70db4affc9aa7f03567a7fa2e2825/tools/jupyter/test/jupyter_notebook_test.py#L10)):*
<details><summary><i>Previously, svenevs (Stephen McDowell) wrote…</i></summary><blockquote>

working, thread block until #21085 lands and can rebase here.
</blockquote></details>

done, after diagnosing, we realized the problem was `pip install --user`.  Removing that, we have the `jupyter` command available and a small cruft removal of an unrelated `jupyter` python package :upside_down_face:


<!-- Sent from Reviewable.io -->

svenevs reviewed (commented):
<img class="emoji" title=":lgtm_cancel:" alt=":lgtm_cancel:" align="absmiddle" src="https://reviewable.io/lgtm_cancel.png" height="20" width="61"/> sorry, we don't want this to merge even if CI passes ... removing feature stamp :wink:

> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/21013#-:-NrwQcrfEgQOF1Jgfuzw:b-5hjpxv)* status: LGTM missing from assignee svenevs



<!-- Sent from Reviewable.io -->

jwnimmer-tri reviewed (commented):
> Reviewed 1 of 1 files at r5, 2 of 3 files at r8, 3 of 3 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/21013)* status: LGTM missing from assignee svenevs



<!-- Sent from Reviewable.io -->

MarcusSorealheis commented:
Thanks everyone for all the help here. A lot of big projects lose the vitality that the cutting edge ecosystem offers because they do not upgrade Python regularly. What a journey.

MarcusSorealheis commented:
Is the next bit of work here @jwnimmer-tri to flash the images? In that case, I would assume the PR will be on pause until then.

jwnimmer-tri commented:
Yup.  The kitware folks need to prepare and deploy the Provisioned images, at which point we can concurrently merge this PR and switch Provisioned CI to use those image.

svenevs reviewed (commented):
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> 

> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/21013#-:-NsGit7bDDPxcRDWrnSn:b-uwkowd)* status: <span>:shipit:</span> complete! all discussions resolved, LGTM from assignees jwnimmer-tri(platform),svenevs



<!-- Sent from Reviewable.io -->

svenevs commented:
> A lot of big projects lose the vitality that the cutting edge ecosystem offers because they do not upgrade Python regularly. What a journey.

You'll have to blame @jwnimmer-tri :slightly_smiling_face:  Drake evolves quickly, while at the same time maintaining stability.  That's a hard balance!

MarcusSorealheis commented:
I've added a PR to not break system packages going forward: https://github.com/RobotLocomotion/drake/pull/21013

MarcusSorealheis commented:
I've added a PR to stop breaking system Python packages, which was introduced by this PR: https://github.com/RobotLocomotion/drake/pull/21269
